src/appendbuf.rs

```diff
@@ -0,0 +1,460 @@
+    /// This guarantees that the AROIobuf can be thought of as safely immutable while this
+    /// buffer can continue to be filled and poked. There are no operations for this buffer
+    /// to reset the window to a lower position in the buffer.
+    /// len is the number of bytes back from the start of the window where the slice begins
+    ///  (and also the length of the slice)
+    ///
+    /// This function does not mutate the buffer, window, or limits.
+    /// It just increases the refcount.
+    ///
+    /// ```rust
+    /// #![allow(unstable)]
+    /// use iobuf::{AppendBuf, Iobuf};
+    ///
+    ///   let mut buf = AppendBuf::new(24);
+    ///   for i in b'A' .. b'X' + 1 {
+    ///     buf.fill_be(i).unwrap();
+    ///   }
+    ///
+    ///   let begin = buf.atomic_slice_from_begin(8).ok().expect("from_begin");
+    ///   let middle = buf.atomic_slice_pos_from_end(16, 8).ok().expect("pos_from_end");
+    ///   let end = buf.atomic_slice_from_end(8).ok().expect("from_end");
+    ///   let meh = buf.atomic_slice_pos_from_begin(4, 8).ok().expect("pos_from_begin");
+    ///
+    ///   let b = unsafe { begin.as_window_slice() };
+    ///   let m = unsafe { middle.as_window_slice() };
+    ///   let e = unsafe { end.as_window_slice() };
+    ///   let z = unsafe { meh.as_window_slice() };
+    ///
+    ///   assert_eq!(b, ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H'].iter().map(|&: x| *x as u8).collect::<Vec<u8>>());
+    ///   assert_eq!(m, ['I', 'J', 'K', 'L', 'M', 'N', 'O', 'P'].iter().map(|&: x| *x as u8).collect::<Vec<u8>>());
+    ///   assert_eq!(e, ['Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X'].iter().map(|&: x| *x as u8).collect::<Vec<u8>>());
+    ///   assert_eq!(z, ['E', 'F', 'G', 'H', 'I', 'J', 'K', 'L'].iter().map(|&: x| *x as u8).collect::<Vec<u8>>());
+    /// ```
+    ///
+    #[inline(always)]
+    pub fn atomic_slice_from_end(&self, len: u32) -> Result<AROIobuf, ()> { unsafe {
+        let mut ret = self.raw.clone_atomic();
+        let lim = (self.raw.lo() - len,self.raw.lo()) ;
+        try!(ret.expand_limits_and_window(lim, lim));
+        Ok(mem::transmute(ret))
+    }}
+
+    /// Creates an AROIobuf as a slice of written buffer. This is space that preceeds
+    /// the window in the buffer, or, more specifically, between the lo_min and lo offsets.
+    /// This guarantees that the AROIobuf can be thought of as safely immutable while this
+    /// buffer can continue to be filled and poked. There are no operations for this buffer
+    /// to reset the window to a lower position in the buffer.
+    /// pos is the offset backwards from the beginning of the window
+    /// len is the length of the slice
+    ///
+    /// This function does not mutate the buffer, window, or limits.
+    /// It just increases the refcount.
+    ///
+    /// ```rust
+    /// #![allow(unstable)]
+    /// use iobuf::{AppendBuf, Iobuf};
+    ///
+    ///   let mut buf = AppendBuf::new(24);
+    ///   for i in b'A' .. b'X' + 1 {
```

Comment:
I should have paid more attention. It turns out there is no inclusive range operator :/